nomos-services/data-availability/sampling/src/backend/kzgrs.rs

```diff
@@ -1,2 +1,397 @@
+use nomos_core::da::BlobId;
+use nomos_da_network_core::SubnetworkId;
+
+const NON_TRIGGERED_ERR: &str =
+    "We should not receive a sampling success from a non triggered blobId";
+
+#[derive(Clone)]
+pub struct SamplingContext {
+    subnets: HashSet<SubnetworkId>,
+    started: Instant,
+}
+
+#[derive(Debug, Clone)]
+pub struct KzgrsDaSamplerSettings {
+    pub num_samples: u16,
+    pub old_blobs_check_interval: Duration,
+    pub blobs_validity_duration: Duration,
+}
+
+pub struct KzgrsDaSampler<R: Rng> {
+    settings: KzgrsDaSamplerSettings,
+    validated_blobs: BTreeSet<BlobId>,
+    pending_sampling_blobs: HashMap<BlobId, SamplingContext>,
+    rng: R,
+}
+
+impl<R: Rng> KzgrsDaSampler<R> {
+    fn prune_by_time(&mut self) {
+        self.pending_sampling_blobs.retain(|_blob_id, context| {
+            context.started.elapsed() < self.settings.blobs_validity_duration
+        });
+    }
+}
+
+// TODO there is no logging at all here. Should we not do some logging?
+// Or can we assume that it is done on an upeer level by the clients?
+#[async_trait::async_trait]
+impl<R: Rng + Sync + Send> DaSamplingServiceBackend<R> for KzgrsDaSampler<R> {
+    type Settings = KzgrsDaSamplerSettings;
+    type BlobId = BlobId;
+    type Blob = DaBlob;
+
+    fn new(settings: Self::Settings, rng: R) -> Self {
+        let bt: BTreeSet<BlobId> = BTreeSet::new();
+        Self {
+            settings,
+            validated_blobs: bt,
+            pending_sampling_blobs: HashMap::new(),
+            rng,
+        }
+    }
+
+    async fn next_prune_interval(&self) -> Interval {
+        time::interval(self.settings.old_blobs_check_interval)
+    }
+
+    async fn get_validated_blobs(&self) -> BTreeSet<Self::BlobId> {
+        self.validated_blobs.clone()
+    }
+
+    async fn mark_in_block(&mut self, blobs_ids: &[Self::BlobId]) {
+        for id in blobs_ids {
+            self.pending_sampling_blobs.remove(id);
+            self.validated_blobs.remove(id);
+        }
+    }
+
+    async fn handle_sampling_success(&mut self, blob_id: Self::BlobId, blob: Self::Blob) {
+        if let Some(ctx) = self.pending_sampling_blobs.get_mut(&blob_id) {
+            ctx.subnets.insert(blob.column_idx as SubnetworkId);
+
+            println!("{}", self.settings.num_samples);
+            println!("{}", ctx.subnets.len());
+            match ctx.subnets.len() {
+                // sampling of this blob_id terminated successfully
+                len if len == self.settings.num_samples as usize => {
+                    self.validated_blobs.insert(blob_id);
+                    // cleanup from pending samplings
+                    self.pending_sampling_blobs.remove(&blob_id);
+                }
+                len if len > self.settings.num_samples as usize => {
+                    unreachable!("{}", "more subnets than expected after sampling success!");
+                }
+                // do nothing if smaller
+                _ => {}
+            }
```

Comment:
```suggestion
        // sampling of this blob_id terminated successfully
        if ctx.subnets.len() == self.settings.num_samples as usize {
            self.validated_blobs.insert(blob_id);
            // cleanup from pending samplings
            self.pending_sampling_blobs.remove(&blob_id);
        }
```

Reply:
I am not much of a fan of hiding away situations which _should_ not happen but if they do, they are critical if not fatal...why would we do that? Yes, the code is nicer. Is there any place in the rest of the codebase which makes this error impossible?

Reply:
As we discussed privately. It is not possible it is more than the expected threshold. We increment 1 by 1 (ensured by `&muy self` when we have enough the queue is empty and the blob is considered validated. Something else may happen. But not that.